sphinx/ext/coverage.py

```diff
@@ -65,6 +66,45 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
     yield _add_line(col_widths, separator)
 
 
+def load_modules(mod_name: str, ignored_module_exps: list[re.Pattern[str]]) -> set[str]:
+    """Recursively load all submodules.
+
+    :param mod_name: The name of a module to load submodules for.
+    :param ignored_module_exps: A list of regexes for modules to ignore.
+    :returns: A set of modules names including the provided module name,
+        ``mod_name``
+    :raises ImportError: If the module indicated by ``mod_name`` could not be
+        loaded.
+    """
+    modules: set[str] = set()
+
+    for exp in ignored_module_exps:
+        if exp.match(mod_name):
+            return modules
+
+    # this can raise an exception but it's the responsibility of the caller to
+    # handle this
+    mod = import_module(mod_name)
+
+    modules.add(mod_name)
+
+    for sub_mod_info in pkgutil.iter_modules(mod.__path__):
+        if sub_mod_info.name == '__main__':
+            continue
+
+        # TODO: Do we want to skip private modules (note: different from private
+        # objects)?
+        # if sub_mod_info.name.startswith('_'):
+        #     continue
+
+        if sub_mod_info.ispkg:
+            modules |= load_modules(f'{mod_name}.{sub_mod_info.name}', ignored_module_exps)
+        else:
+            modules.add(f'{mod_name}.{sub_mod_info.name}')
```

Comment:
Don't you want to exclude a submodule if the latter matches a regexpr?

Reply:
Good point. Done.